torch/utils/cpp_extension.py

```diff
@@ -917,6 +919,42 @@ def _prepare_ldflags(extra_ldflags, with_cuda, verbose):
     return extra_ldflags
 
 
+def _get_cuda_arch_flags(cflags=None):
+    # If cflags is given, there may already be user-provided arch flags in it
+    # (from `extra_compile_args`)
+    if cflags is not None:
+        for flag in cflags:
+            if 'arch' in flag:
+                return []
+
+    # The default is sm_30 for CUDA 9.x and 10.x
+    # First check for an env var (same as used by the main setup.py)
+    # Can be one or more architectures, e.g. '6.1' or "3.5 5.2 6.0 6.1 7.0+PTX"
+    # See cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake
+    arch_list = os.environ.get('TORCH_CUDA_ARCH_LIST', None)
```

Comment:
1. `TORCH_CUDA_ARCH_LIST` is commonly used with semi-colon in-between options, like here: https://github.com/pytorch/builder/blob/master/conda/pytorch-1.1.0/build.sh#L32 , so you have to also split on that. In fact, I've never seen it be split on space, I think I'm seeing it first in the comments of select_compute_arch.cmake 
2. PTX can be added to any architecture flag, not just to 7.0 / 7.2 / 7.5. For example, 3.5+PTX is totally a thing. Basically, when you add PTX, it also adds the PTX version of the flags (I have to pull up from memory how they look).
3. When something like `Pascal` is asked, then you have to include `6.0;6.1`, not just the lower one. This is important, or else there will be speed loss in the compiled kernels on particular GPUs.

Reply:
Thanks for the insights @soumith.

 > In fact, I've never seen it be split on space, I think I'm seeing it first in the comments of `select_compute_arch.cmake`

Rather than accommodate both, I could just change that comment to use `;` and only allow that? Keeps the logic simpler; allowing too many combinations may be fragile. If you've never seen it, then it's likely no one is using it.

> Basically, when you add PTX, it also adds the PTX version of the flags (I have to pull up from memory how they look)

PTX is `compute_xx`. In `-gencode=arch=X,code=Y`, `X` is always `compute_##`, while `Y` can be either `sm_##`, `compute_##` (or even both). Doing all that via the env var is going to be really complicated. That's why I proposed to always use `-gencode=arch=compute_xx,code=sm_xx`, and if a user wants any other flavors she can simply pass those in via `extra_cuda_cflags` directly.

Or maybe the one extra flavor to add is `-gencode=arch=compute_xx,code=compute_xx` for `"xx+PTX"` - is that what you meant?

Reply:
> Rather than accommodate both, I could just change that comment to use ; and only allow that?

Yea that sounds good.

> Or maybe the one extra flavor to add is -gencode=arch=compute_xx,code=compute_xx for "xx+PTX" - is that what you meant?

Yes, I meant that the `+PTX` has to be generically parsed out for all architectures, and it seems like it's not hard to do.

Reply:
All comments addressed.

> Rather than accommodate both, I could just change that comment to use ; and only allow that?

I found another example with spaces, in `docker/pytorch/Dockerfile`, so supported both delimiters after all.